src/display/annotation_layer.js

```diff
@@ -476,6 +478,35 @@ class TextWidgetAnnotationElement extends WidgetAnnotationElement {
         event.target.setSelectionRange(0, 0);
       });
 
+      if (this.data.actions) {
+        element.addEventListener("updateFromSandbox", function (event) {
+          const data = event.detail;
+          if ("value" in data) {
+            event.target.value = event.detail.value;
+          } else if ("focus" in data) {
+            event.target.focus({ preventScroll: false });
+          }
+        });
+
+        for (const eventType of Object.keys(this.data.actions)) {
+          switch (eventType) {
+            case "Format":
+              element.addEventListener("blur", function (event) {
+                window.dispatchEvent(
+                  new CustomEvent("dispatchEventInSandbox", {
```

Comment:
I see how the sandbox gets the ID, name and value, but where does the sandbox itself get the other formatting parameters from (like the number of decimals)?

Reply:
So all the fields are got through `getFieldObjects` and then sent to `code_gen.js`:
https://github.com/mozilla/pdf.js/pull/12432/files#diff-a4a4f7fb5d721f02823c9aabe4eb6c6caf33465f75a88f5e823d4c2851f26bf8R106
All the generated code here is `eval` in the sandbox.
At the end we'll have a mapping id->Field, so using this map we can run actions (the Format ones) in the sandbox.
The Format action can be `AFNumber_Format(2,2,0,0,"",false )`, in this case after the action has ran (and transformed the number into a 2 digits one), the `source.value` is modified with the new value (2 digits) (https://github.com/mozilla/pdf.js/pull/12432/files#diff-388aefce79f87599ad89abb4759ed56c669e2e53c21ee1889b11f223ff7a524aR63) and when setting this value the proxy will then send a message with the new value (the 2 digits one) and this message will then arrive here:
https://github.com/mozilla/pdf.js/pull/12432/files#diff-e24e4bd32090b0c400a70f02d1b844df95c9bbe62cd0b280f4397e9f4194fe44R1354
and...
Anyway for my local dev, I use quickjs as a sandbox (a way faster to use locally):
https://github.com/calixteman/pdf.js/blob/sandbox_experimental/src/scripting_api/sandbox.js
and the code for the quickjs part:
https://github.com/calixteman/pdf.js.quickjs
Do you I make a PR with it ? at least to test how all that stuff is working.